### ISSUE TITLE
Re-evaluate security implication of webcomponents

### ISSUE THREAD

**Issue by glacambre** [tridactyl/tridactyl#1134]
Webcomponents landed in Firefox 63: https://hacks.mozilla.org/?p=32854

We need to make sure that the addEventListener/Focus hijacking code is not negatively impacted by them.

I'll have time to do that tomorrow.

**Comment by bovine3dom** [comment]
Can we use the shadow DOM instead of cleanslate?

**Comment by glacambre** [comment]
Ah, yes, it might even solve the issues `noiframeon` works around.

**Comment by glacambre** [comment]
Let's recapitulate our worries: we're afraid that our [eventListener-hijacking](https://github.com/tridactyl/tridactyl/blob/de9fc2d4b5ba10e8ed2da109e72a491369fb21ef/src/lib/dom.ts#L453) code might let attackers execute code in the extension context. We though it might happen because we [expose](https://github.com/tridactyl/tridactyl/blob/de9fc2d4b5ba10e8ed2da109e72a491369fb21ef/src/lib/dom.ts#L455) a function, [registerEvListenerAction](https://github.com/tridactyl/tridactyl/blob/de9fc2d4b5ba10e8ed2da109e72a491369fb21ef/src/lib/dom.ts#L404), to the page. If a malicious element overrides one of the functions used by Tridactyl, would this function be called? Would it have extension privileges? The MDN documentation tells us [no](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Xray_vision).

In order to make sure of that, I wrote the following page:
```
<!DOCTYPE html>
<html>
        <head>
                <meta charset="utf-8">
                <title></title>
                <script>
                        class EvilElement extends HTMLParagraphElement {
                                constructor() {
                                        super()
                                }
                                connectedCallback() {
                                        console.log("Evil element inserted in page.");
                                }
                                addEventListener() {
                                        console.log("Downloading malware.")
                                }
                                getClientRect() {
                                        console.log("Executing malware.")
                                }
                        }
                        customElements.define('evil-element', EvilElement, { extends: "p" });
                        addEventListener("load", e => {
                                EventTarget.prototype.addEventListener.call(document.getElementById("evil"), "click", () => console.log(tri))
                        })
                </script>
        </head>
        <body>
                <p id="evil" is="evil-element">Hello</p>
        </body>
</html>
```
The EvilElement overrides the default getClientRect() function elements have. This is a function Tridactyl [uses](https://github.com/tridactyl/tridactyl/blob/de9fc2d4b5ba10e8ed2da109e72a491369fb21ef/src/content/hinting.ts#L267) when highlighting elements for hinting. When opening the page, the `Evil element inserted in page.` message is displayed in the console, so we know it works. When pressing `f`, the evil element is correctly highlighted, but no messages are logged in the console, so we know the original `getClientRect()` was called instead of the user defined one. When clicking on the link, `undefined` is logged instead of the `tri` element, so we know the callback is executed in the page context instead of the extension one.

In conclusion, Tridactyl is not affected by webcomponents.
Related PR: https://github.com/tridactyl/tridactyl/pull/225.
